crates/agent-sql/src/live_specs.rs

```diff
@@ -0,0 +1,162 @@
+pub struct LiveSpec {
+    pub id: Id,
+    pub last_pub_id: Id,
+    pub last_build_id: Id,
+    pub catalog_name: String,
+    pub spec_type: Option<CatalogType>,
+    pub spec: Option<TextJson<Box<RawValue>>>,
+    pub built_spec: Option<Json<Box<RawValue>>>,
+    pub inferred_schema_md5: Option<String>,
+    // User's capability to the specification `catalog_name`.
+    pub user_capability: Option<Capability>,
+    // Capabilities of the specification with respect to other roles.
+    pub spec_capabilities: Json<Vec<RoleGrant>>,
+}
+
+/// Returns a `LiveSpec` row for each of the given `names`. This will always return a row for each
+/// name, even if no live spec exists in the database.
+pub async fn fetch_live_specs(
+    user_id: Uuid,
+    names: &[String],
+    db: impl sqlx::Executor<'_, Database = sqlx::Postgres>,
+) -> sqlx::Result<Vec<LiveSpec>> {
+    sqlx::query_as!(
+        LiveSpec,
+        r#"
+        select
+            coalesce(ls.id, '00:00:00:00:00:00:00:00'::flowid) as "id!: Id",
+            coalesce(ls.last_pub_id, '00:00:00:00:00:00:00:00'::flowid) as "last_pub_id!: Id",
+            coalesce(ls.last_build_id, '00:00:00:00:00:00:00:00'::flowid) as "last_build_id!: Id",
+            names as "catalog_name!: String",
+            ls.spec_type as "spec_type?: CatalogType",
+            ls.spec as "spec: TextJson<Box<RawValue>>",
+            ls.built_spec as "built_spec: Json<Box<RawValue>>",
+            ls.inferred_schema_md5,
+            (
+                select max(capability) from internal.user_roles($1) r
+                where starts_with(names, r.role_prefix)
+            ) as "user_capability: Capability",
+            coalesce(
+                (select json_agg(row_to_json(role_grants))
+                from role_grants
+                where starts_with(names, subject_role)),
+                '[]'
+            ) as "spec_capabilities!: Json<Vec<RoleGrant>>"
+        from unnest($2::text[]) names
+        left outer join live_specs ls on ls.catalog_name = names
+        "#,
+        user_id,
+        names,
+    )
+    .fetch_all(db)
+    .await
+}
+
+pub struct InferredSchemaRow {
+    pub collection_name: String,
+    pub schema: Json<Box<RawValue>>,
+    pub md5: String,
+}
+
+pub async fn fetch_inferred_schemas(
+    collections: &[&str],
+    pool: &sqlx::PgPool,
+) -> sqlx::Result<Vec<InferredSchemaRow>> {
+    sqlx::query_as!(
+        InferredSchemaRow,
+        r#"select
+            collection_name,
+            schema as "schema!: Json<Box<RawValue>>",
+            md5 as "md5!: String"
+            from inferred_schemas
+            where collection_name = ANY($1::text[])
+            "#,
+        collections as &[&str],
+    )
+    .fetch_all(pool)
+    .await
+}
+
+/// Queries for all non-deleted `live_specs` that are connected to the given `collection_names` via
+/// `live_spec_flows`.
+pub async fn fetch_expanded_live_specs(
```

Comment:
This routine is run only on user-initiated publications, as I understand :+1:

One note is we still need the recursive resolution behavior of resolve_expanded_rows() as well, but now we only need it to pull in live and unchanged derivations which are attached to any drafted tests.

We still need to recursively walk such derivations to pull out the full data flow cycle being put under test.

Reply:
I wasn't thinking we needed to do that. The expansion logic here applies to the draft, which will subsequently go though another round of resolution to pull in all the collections referenced by tasks/tests in the draft. So this function expands from published collections to any tests/tasks that depend on them, and then the subsequent `live_specs` resolution will pull in collection specs that those tests need in order to run. I really should have more explicit tests for this, which might help to illustrate that better. But I'm not 100% sure I've got your point. Are you saying that we should try to expand through an arbitrary number of derivations until we get to the captures/materializations "edges"? _That_ would be pretty different from what's here.

Reply:
BTW the best spec expansion test we have right now is [`user_publications.rs`](https://github.com/estuary/flow/blob/2570da4f8b7b028da9f45f058159189f2b06377e/crates/agent/src/integration_tests/user_publications.rs#L167).

Reply:
We discussed this on VC and I created #1494 to track the issue